hbase_storage/src/main/java/com/splicemachine/si/impl/region/HBaseTxnFinder.java

```diff
@@ -0,0 +1,101 @@
+package com.splicemachine.si.impl.region;
+
+import com.splicemachine.hbase.CellUtils;
+import com.splicemachine.si.impl.TxnUtils;
+import com.splicemachine.utils.Pair;
+import com.splicemachine.utils.Source;
+import org.apache.hadoop.hbase.Cell;
+import org.apache.hadoop.hbase.client.Scan;
+import org.apache.hadoop.hbase.filter.PrefixFilter;
+import org.apache.hadoop.hbase.regionserver.HRegion;
+import org.apache.hadoop.hbase.regionserver.RegionScanner;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+public class HBaseTxnFinder implements TxnFinder {
+
+    private HRegion region;
+
+    public HBaseTxnFinder(HRegion region) {
+        this.region = region;
+    }
+
+    private class ScanTimestampIterator implements Source<Pair<Long, Long>> {
+        private final RegionScanner regionScanner;
+        protected Pair<Long, Long> next;
+        private List<Cell> currentResults;
+
+        protected Pair<Long, Long> decode(List<Cell> keyValues) throws IOException{
+            if(keyValues.size()<=0) return null;
+            Cell dataKv=null;
+            long ts = 0;
+
+            for(Cell kv : keyValues){
+                if(CellUtils.singleMatchingColumn(kv,V2TxnDecoder.FAMILY,V2TxnDecoder.DATA_QUALIFIER_BYTES)) {
+                    dataKv=kv;
+                }
+                if(ts <  kv.getTimestamp()) {
+                    ts = kv.getTimestamp(); // get last update
+                }
+            }
+            if(dataKv==null) return null;
+            long txnId= TxnUtils.txnIdFromRowKey(dataKv.getRowArray(),dataKv.getRowOffset(),dataKv.getRowLength());
```

Comment:
If I'm understanding this correctly, this is mapping the Txn begin timestamp to the instant when the transaction was last updated. This would be wrong for long running transactions:
1. txn1 is created at instant1
2. a long time passes
3. txn1 is committed at instant500
5. this function maps txnId 1 to instant 500 (it should map txnId 1 to instant 1)

Reply:
You're right, what's a better way to correlate? maybe parse and check the timestamp of `COMMIT_QUALIFIER_BYTES` or `STATE_QUALIFIER_BYTES` (sorry maybe we already went through this once, but I just want to make sure).

Reply:
I think the best option would be to scan for STATE_QUALIFIER_BYTES with ACTIVE value (0x00), since that's what's written when the transaction is first created. In principle there should only be 2 states at most (ACTIVE, superseded by COMMITTED or ROLLEDBACK) but I think I saw some cases where ROLLEDBACK was written multiple times. If there's such a case and the ACTIVE keyvalue is not available I'd ignore that row.